[FEAT] 로그인 여부 반환 API 추가

jye326 opened the pull request:
## Issue Number
closed #1151 

## As-Is
<!-- 문제 상황 정의 -->
accessToken을 소유한 클라이언트가 자신의 memberId를 확인할 수 없었습니다. 소셜 로그인의 경우 리다이랙트 요청이라 직접 memberId를  응답해 줄 수 없고, accessToken만을 응답 쿠키에 포함할 수 있는데 이 경우 클라이언트에서 memberId를 획득할 수단이 필요했습니다.

## To-Be
<!-- 변경 사항 -->
- `/isLoggedIn` 엔드포인트 추가
  - accessToken 존재 여부와 로그인의 상태 반환
  - 쿠키에서 accessToken을 이용해 멤버 ID 추출
- `LoginStatusDto` 추가
  - 로그인 상태를 나타내는 DTO 생성
- 기존 코드 정리 및 리팩토링
  - 불필요한 import와 jwt validation 중복 로직 제거

## 로그인 여부 확인 API 설계 결정 과정
### 1. 응답 위치 결정: Interceptor vs Controller

- 고민 : 인터셉터에서 토큰을 검증한 즉시 응답을 반환(Short-circuit)하여 성능을 최적화할지 고민
- 판단 : 직접 HttpServletResponse에 JSON을 작성하는 방식은 유지보수 난이도를 높이고, 전역 에러 핸들러나 Swagger 문서 자동화 등 프레임워크의 이점을 누리기 어려움. 얻는 성능 이점에 비해 잃는 관리 편의성이 더 크다고 판단함.
- 결정 : Controller에서 반환하는 기존 방식과 양식 통일 결정

### 2. 인증 필수 여부에 따른 로직 충돌
- 고민 : 기존에 인터셉터에서 @AuthRequired 어노테이션을 기준으로 토큰이 없으면 즉시 예외(Unauthorized)를 던지는 로직이 있어 이를 활용하여 중복 코드를 최소화 할까 고민.
- 판단 : isLoggedIn API는 토큰이 없어도 예외가 아닌 false를 반환해야 하므로, @AuthRequired에 required=false 옵션을 추가해 분기 처리를 할까 고민함.

### 3. 어노테이션의 의미적 모순(Semantics) 발견
- 문제: @AuthRequired라는 명칭 자체가 "인증이 필수임"을 뜻함에도 불구하고, 내부 옵션으로 required=false를 주는 것은 코드의 가독성과 직관성을 해치는 모순적인 설계임.
- 결론: 해당 API에는 @AuthRequired를 붙이지 않기로 함.

### 4. 최종 구현 방식: 컨트롤러/서비스 레이어 처리
- 결정: 다소의 코드 중복을 허용하더라도 컨트롤러 단 이하에서 로직을 처리함.

구현 방향:

- 인터셉터는 **인증이 강제되는 API**만 담당하도록 역할을 명확히 제한함.
- isLoggedIn과 같이 인증이 선택적인 API는 컨트롤러(또는 서비스)에서 직접 쿠키/토큰을 확인하는 비즈니스 로직을 수행함.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

jye326 committed:
feat: 로그인 상태 확인 및 쿠키 기반 멤버 ID 추출 기능 추가

- `/isLoggedIn` 엔드포인트 추가
  - accessToken 존재 여부와 로그인의 상태 반환
- `extractMemberId` 메서드 추가
  - 쿠키에서 accessToken을 이용해 멤버 ID 추출
- `LoginStatusDto` 추가
  - 로그인 상태를 나타내는 DTO 생성
- 기존 코드 정리 및 리팩토링
  - 불필요한 import와 jwt validation 중복 로직 제거

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -72,6 +84,14 @@ public ResponseEntity<Void> logout(@Login LoginInfo loginInfo, HttpServletRespon
+        Long memberId = authService.extractMemberId(cookies);
+        return ResponseEntity.status(HttpStatus.OK)
+                .body(new LoginStatusDto(memberId != null, memberId));
+    }
```
p1: API의 url과 응답에 대해서 다시 한번 생각해볼 필요가 있을 것 같아요. 
첫번째로 Restful API 는 엔드포인트를 동사보단 명사로 표현하는 것을 추천하고 있습니다. 
동사보단 `auth/me` 같은 명사로 나타는 것이 어떨까요? 

두번째로 응답을 boolean보단 상태코드로 하는 것은 어떨까요?
현재 인증상태 판별은 상태코드로 충분히 할 수 있을 것 같아요. 클라이언트 입장에서도 응답을 true/false 보단 인증 되었다면 `200 OK + 응답 body`, 실패했다면 `204 No Content` 로 나타내면 더욱 좋을 것 같아요!

jye326 replied:
1. `auth/check`로 변경했습니다.
2. 말씀하신 대로 204가 적절해 보여 변경하였습니다.

kimdevv commented on backend/fittoring/src/main/java/fittoring/application/auth/service/AuthService.java:
```diff
@@ -163,4 +165,17 @@ private Member getRandomIdPwMember(OauthSignUpRequest request) {
+
+    public Long extractMemberId(Cookie[] cookies) {
+        // 쿠키가 아예 없는 경우, 첫 접속 or 비회원
+        if(cookies==null||cookies.length==0){
```
p1: 띄어쓰기가 들어가면 읽기 편할 것 같아요!

jye326 replied:
반영했습니다

kimdevv commented on backend/fittoring/src/main/java/fittoring/application/auth/service/AuthService.java:
```diff
@@ -163,4 +165,17 @@ private Member getRandomIdPwMember(OauthSignUpRequest request) {
+            String accessToken = jwtExtractor.extractTokenFromCookie("accessToken", cookies);
+            return jwtProvider.getSubjectFromPayloadBy(accessToken);
+        } catch (Exception e) {
+            return null;
```
p1: Jwt를 검증하거나 추출하는 과정에서 예외가 발생했을 때에는 null을 반환하는 것보다는 커스텀 예외를 던져주는 게 디버깅하기 편할 것 같아요! `memberId`가 null인 경우 **쿠키 자체가 비어 있는 상황**이었는지, 아니면 **Jwt 관련 예외가 터진 상황**이었는지 모호할 것 같습니다.

jye326 replied:
쿠키 배열이 비어있거나 accessToken 쿠키가 없을 경우는 null을 반환하여 204 정상 응답을 보내고, 검증 및 추출 오류인 경우에는 Exception을 던지도록 변경했습니다.

kimdevv reviewed (commented):
고생하셨습니다~ 몇 가지 커멘트 남겨놨습니다!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/service/AuthService.java:
```diff
@@ -163,4 +165,17 @@ private Member getRandomIdPwMember(OauthSignUpRequest request) {
+        }
+        try {
+            String accessToken = jwtExtractor.extractTokenFromCookie("accessToken", cookies);
+            return jwtProvider.getSubjectFromPayloadBy(accessToken);
```
p1: 여기서 예외가 발생한 경우 null로 예외를 던지고 있어요. 만약 AT가 만료된 경우 재발급을 받을 수 있도록 401 응답이 반환되어야 하는데 현재는 그러지 못하는 것 같아요!

jye326 replied:
만료의 경우 예외를 그대로 던지도록 변경했습니다!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/service/AuthService.java:
```diff
@@ -163,4 +165,17 @@ private Member getRandomIdPwMember(OauthSignUpRequest request) {
     }
+
+    public Long extractMemberId(Cookie[] cookies) {
+        // 쿠키가 아예 없는 경우, 첫 접속 or 비회원
```
p2: 쿠키를 여기서 사용하면 테스트시 HttpRequest 객체가 필요해 Service 테스트가 어려울 것 같아요. Cookie를 서비스까지 보내지 말고 컨트롤러에서 검증하면 어떨까요?

jye326 replied:
쿠키의 추출을 컨트롤러에서 하게 되면 관련 예외 처리 로직이 필요하여 일단 서비스 단에 유지했습니다! 쿠키에서 값 추출에 실패한 경우 null을 통한 204응답을 보내고, 만료의 경우는 예외를 던지도록 했습니다~

dye0p reviewed (changes requested):
수고하셨습니다! 몇가지 코멘트 남겼습니다~ 확인 부탁드려요!

jye326 committed:
refactor: AuthController 로그인 체크 로직 및 DTO 개선

- `extractMemberId()` 메서드의 쿠키 의존성을 제거하고 accessToken 기반으로 변경
- `/isLoggedIn` 엔드포인트를 `/auth/check`로 리네이밍
- `LoginStatusDto`에서 `isLoggedIn` 필드 제거 및 `memberId`로 대체
- 예외 처리 및 상태 코드 일관성 개선 (204 상태 코드 추가)

dye0p reviewed (approved):
의견 반영이 잘 된 것 같아서 이만 승인합니다~
